waspc/src/Wasp/Project/Analyze.hs

```diff
@@ -82,9 +85,15 @@ constructAppSpec waspDir options decls = do
             AS.devDatabaseUrl = devDbUrl,
             AS.customViteConfigPath = customViteConfigPath
           }
-  return $ case validateAppSpec appSpec of
-    [] -> Right appSpec
-    validationErrors -> Left $ map show validationErrors
+  let (validationErrors, validationWarnings) =
+        let errsAndWarns = validateAppSpec appSpec
+         in ( filter isValidationError errsAndWarns,
+              filter isValidationWarning errsAndWarns
+            )
+  return
+    ( if null validationErrors then Right appSpec else Left (show <$> validationErrors),
```

Comment:
TLDR; @sodic I agree that we don't agree on this one, and while I don't want to stop writing like this because I find it to be superior, I also won't force you to use my style because it is not important enough.

As for the discussion, I think that old discussion @sodic linked quite well captures the arguments from my side.

For me, at this point, it is easier to read and reason with `<$>` and `<>` then with `map` and `++`. It wasn't like that at some point in the past, but once I got into it, I find it better and easier to work with. I would suggest to try and sincerely use it for some time before you dismiss it.

As for the argumentation above:

I don't agree with the following statement from @sodic :

> The same rules apply for overly-general type signatures and unnecessary polymorphism.

This is not the same thing as what we are discussing.

> My rule of thumb is: It's always best to go for the most specific (semantically meaningful) function for a task.

Also don't agree with this. I find it quite beneficial now to have to reason about the types here, I just care about what `<$>` and `<>` do. I did the refactoring just last week where if I had used `++` I would have had to convert them all into `<>` or something else, but I didn't have to because I already used `<>`, so it saved me some extra work.

> We wouldn't write:
...

Also don't agree: I wouldn't write the first one, I will sometimes do the second one, and I will almost always write the third one.

> In fact, I've been giving [the same advice](https://puh.takelab.fer.hr/puh-23-24-assignments/petar-dusevic/-/merge_requests/9#note_5319) to students in our course.

One, that is a bit different situation and I wouldn't use it as an argument for this discussion, and two, the fact that you are giving that advice to others doesn't make it correct.